Challenge 6 multisig init

rin-st opened the pull request:
Currently you need to run additionally `yarn backend-local` (express server) to save transactions. Currently I still need to fix todos below, but it's minor things and you can start to test the challenge  without these.

TODOs
- [x] complete transfer function
- [x] update styles
- [x] update Readme @Pabl0cks 
- [x] refactor to hooks
- [x] move backend-local to nextjs api for networks other than hh
- [x] update main page and metadata @Pabl0cks 
- [x] **Completed** text instead of buttons for comleted transactions
- [x] recheck // TODO's in code

Pabl0cks reviewed (commented):
I was getting some trouble executing a transaction with more than the required signers, the button was turning secondary style and I was getting this error, having no duplicated signatures: "executeTransaction: duplicate or unordered signatures"

Just pushed a small tweak to try to fix it in 37335d716bcfd1610961bed1441331f9fdb7e9c4

Pabl0cks commented:
Awesome job @rin-st! First time I try a multisig and it has been really smooth and really like your implementation 🙌

I've noted some improvements, behaviors to review, and possible bugs.

- [ ] If you select "transferFunds" in the Create tab, and then try to do an addSigner from Owners, when clicking "Create TX" you get an error: AbiFunctionNotFoundError: Function "transferFunds" not found on ABI. Should we update local storage from owners?
- [ ] Events in Multisig tab => Amounts are shown with $ symbol instead of ETH
- [ ] Add loading to Pool tab while data is loading
- [ ] Check if amount of required signatures is higher than Current signers -1 in case of removeSigner
- [ ] It takes a couple of seconds to load Pools tab, we can add a loading/spinner 
- [ ] Should we give some feedback if an address who is not a signer tries to propose a transaction?
- [ ] Should I be able to try to send an AddSigner directly from Create tab? => Calldata is missing, I'm getting an error. 
- [ ] Number of required signs should not get updated for old transactions in pool. I've changed from 2 to 1, and old transaction shows 2/1
- [ ] Disable Exec when there isn't enough signers
- [ ] Disable Sign if you already signed that transaction

rin-st commented:
Good points, thanks!

> If you select "transferFunds" in the Create tab, and then try to do an addSigner from Owners, when clicking "Create TX" you get an error: AbiFunctionNotFoundError: Function "transferFunds" not found on ABI. Should we update local storage from owners?

Yes, I think it should be changed to default `addSigner` on page opening.

> Should I be able to try to send an AddSigner directly from Create tab? => Calldata is missing, I'm getting an error.

No, you shouldn't. I think it's ok to change predefined method to `transferFunds` if no calldata, so it will be impossible to choose addSigner. But we should be careful here, since useLocalstorage is used here, on first load calldata can be undefined even if we created add/remove calldata before.

>  Check if amount of required signatures is higher than Current signers -1 in case of removeSigner

why? I think it's ok if I can sign removing me from owners. Or what do you mean?

> Number of required signs should not get updated for old transactions in pool. I've changed from 2 to 1, and old transaction shows 2/1

Agree

rin-st commented:
> I mean this edge case:
Got 2 required signers, and 2 signers registered.
We can exec a removeSigner without updating the required signers to 1.
The multisig becomes useless => 2 required signers and just 1 signer left.

Thanks, got it. I think user should be careful about it, not us. So if he will make that mistake, he will understand more about multisig owners :)

Pabl0cks commented:
> Thanks, got it. I think user should be careful about it, not us. So if he will make that mistake, he will understand more about multisig owners :)

Yeah, and we can always set that check as a side quest.

Pabl0cks commented:
After f7b9442620f1ccc6d8d51fb80272c3f14791e473 I was receiving a runtime error when trying to transferFunds, after input an address and an amount, and clicking "Create". The runtime error was in `getTransactionHash` and error was: _TypeError: Cannot read properties of undefined (reading 'lenght')_
 
Pushing a fix, but not 100% sure if it's correct. The bug stopped but not sure if it's hacky.

Pabl0cks commented:
During the last round of tests and creation of Readme, I've noted a few more items to check:

- [ ] The behaviour when you have nonces that are unexecutables before they're old, is a bit strange. Maybe we should add a new status label to nonces unexecutables, or hide them or set Sign and Exec button to disabled.
- [ ] Should we show nonce 0 after first proposal? Before first proposal, maybe we show something different to "Loading..."? Seems like if we having loading problems.
- [ ] In Multisig tab, should we add a details button too?
- [ ] When navigating manually into Create tab, it's loading again addSigner instead of forcing transferFunds method.

rin-st commented:
Great job @Pabl0cks ! Nice changes! ❤️ 

> The behaviour when you have nonces that are unexecutables before they're old, is a bit strange. Maybe we should add a new status label to nonces unexecutables, or hide them or set Sign and Exec button to disabled.

Added "Outdated" status if nonce is newer than tx.nonce [3c24b03](https://github.com/scaffold-eth/se-2-challenges/pull/112/commits/3c24b038c483d9e9cd6fcdc6b01986588266f812)

> Should we show nonce 0 after first proposal? Before first proposal, maybe we show something different to "Loading..."? Seems like if we having loading problems.

Changed it to contract even when it's 0. Showing `Loading...` only when hook returns `undefined`. [be90072](https://github.com/scaffold-eth/se-2-challenges/pull/112/commits/be90072366b47b1a1500420c6f3224d7f1831ae4)

> In Multisig tab, should we add a details button too?

I'm not sure we need it even in "Pool" tab, since it mostly copies data from table. Added it for consistency with se-1.

> When navigating manually into Create tab, it's loading again addSigner instead of forcing transferFunds method.

fixed [353f074](https://github.com/scaffold-eth/se-2-challenges/pull/112/commits/353f0746cb8fd0cd4389006323259a18c226a107)

Pabl0cks commented:
Great job @rin-st, thanks for tackling everything so quickly! 🙌

> I'm not sure we need it even in "Pool" tab, since it mostly copies data from table. Added it for consistency with sre-1.

You're right, we probably can just skip it, as is just like an event log, and we can check all data in Pool tab 👌

Pabl0cks commented:
I've added a section below the introduction in the README to provide users with more context about this Challenge => It's not about 'filling the missing parts', instead, it serves as an inspirational walk through of the Multisig Wallet features.

ZakGriffith commented:
When sending funds, instead of this address field saying 'Signer address', I think 'Recipient' would make more sense.

![image](https://github.com/scaffold-eth/se-2-challenges/assets/12072395/2ff001de-2708-4f2e-8a4c-0eafdd09b104)

technophile-04 commented on packages/backend-local/index.ts:
fixed at [940c26e](https://github.com/scaffold-eth/se-2-challenges/pull/112/commits/940c26e51b224d76bd0fd1e10d35ecfe923bdb3e), will create an issue for this

technophile-04 commented:
Tysm all !! This is great! tested and I think this looks ready to be merged as the first iteration 🙌

Will create some meta todo's for this challenge and we can iterate their 

---

added try / catch block at [6402b97](https://github.com/scaffold-eth/se-2-challenges/pull/112/commits/6402b97e8b32e3f9a5ab1ecc286a1ab4a3bdbd1c) because I bymistake rejected the transaction and it broke : 

![Screenshot 2023-12-13 at 9 40 26 PM](https://github.com/scaffold-eth/se-2-challenges/assets/80153681/ea0cb447-5038-44c6-a2cc-6203db7966b0)

also did some minor cleanup

technophile-04 reviewed (approved):
Tysm @rin-st and all !! I think this is a good first iteration merge 🙌

The only TODO which is remaining is deploying a PROD server and updating README with that link. 

Just created #117, maybe we can tackle it first and then tell Carlos / Austin to deploy to prod ?
